packages/vite/src/node/plugin.ts

```diff
@@ -121,66 +306,75 @@ export interface Plugin<A = any> extends RollupPlugin<A> {
-   *
-   * - The hook can return a filtered list of modules to narrow down the update.
-   *   e.g. for a Vue SFC, we can narrow down the part to update by comparing
-   *   the descriptors.
-   *
-   * - The hook can also return an empty array and then perform custom updates
-   *   by sending a custom hmr payload via server.hot.send().
-   *
-   * - If the hook doesn't return a value, the hmr update will be performed as
-   *   normal.
+   * Inject per environment plugins after the shared plugin
+   */
+  environmentPlugins?: (
+    environment: PluginEnvironment,
+  ) => PluginOption[] | undefined
+
+  /**
+   * @deprecated
+   * Compat support, ctx.modules is a backward compatible ModuleNode array
+   * with the mixed client and ssr moduleGraph. Use hotUpdate instead
    */
   handleHotUpdate?: ObjectHook<
     (
       this: void,
       ctx: HmrContext,
     ) => Array<ModuleNode> | void | Promise<Array<ModuleNode> | void>
   >
-
-  /**
-   * extend hooks with ssr flag
-   */
-  resolveId?: ObjectHook<
-    (
-      this: PluginContext,
-      source: string,
-      importer: string | undefined,
-      options: {
-        attributes: Record<string, string>
-        custom?: CustomPluginOptions
-        ssr?: boolean
-        /**
-         * @internal
-         */
-        scan?: boolean
-        isEntry: boolean
-      },
-    ) => Promise<ResolveIdResult> | ResolveIdResult
-  >
-  load?: ObjectHook<
-    (
-      this: PluginContext,
-      id: string,
-      options?: { ssr?: boolean },
-    ) => Promise<LoadResult> | LoadResult
-  >
-  transform?: ObjectHook<
-    (
-      this: TransformPluginContext,
-      code: string,
-      id: string,
-      options?: { ssr?: boolean },
-    ) => Promise<TransformResult> | TransformResult
-  >
 }
 
 export type HookHandler<T> = T extends ObjectHook<infer H> ? H : T
 
 export type PluginWithRequiredHook<K extends keyof Plugin> = Plugin & {
   [P in K]: NonNullable<Plugin[P]>
 }
+
+export type MaybeEnvironmentPlugin =
+  | EnvironmentPlugin
+  | false
+  | null
+  | undefined
+
+export type EnvironmentPluginOption =
+  | MaybeEnvironmentPlugin
+  | EnvironmentPluginOption[]
+  | Promise<MaybeEnvironmentPlugin | EnvironmentPluginOption[]>
+
+export type MaybePlugin = Plugin | false | null | undefined
+
+export type PluginOption =
+  | MaybePlugin
+  | PluginOption[]
+  | Promise<MaybePlugin | PluginOption[]>
+
+export async function resolveEnvironmentPlugins(
+  environment: PluginEnvironment,
+): Promise<EnvironmentPlugin[]> {
+  const resolvedPlugins: EnvironmentPlugin[] = []
+  for (const plugin of environment.config.plugins) {
+    if (plugin.environmentPlugins) {
+      const environmentPlugins = await plugin.environmentPlugins(environment)
+      if (environmentPlugins) {
+        const newPlugins =
+          await asyncFlattenEnvironmentPlugins(environmentPlugins)
+        resolvedPlugins.push(...newPlugins)
+      }
+    } else {
+      resolvedPlugins.push(plugin)
+    }
```

Comment:
It looks like `plugin.transform` is dropped when there is `plugin.environmentPlugins`. I brought up the same thing in https://github.com/vitejs/vite/pull/16471#discussion_r1583989466 when it was `create`. Did you forget to do this https://github.com/vitejs/vite/pull/16471/commits/12d467f642a0d71fcc718d036d20dc4b6efda549 or intentional?

Reply:
This is unintentional

Reply:
Thanks @hi-ogawa! This should be fixed at https://github.com/vitejs/vite/pull/16471/commits/63de43a5cda8fbab73e3487a168161d4172ef31e. We really need tests for the new `environmentPlugins`.

Interested in your feedback about the hook vs functional form now that you have used both.

Reply:
Thanks for the fix!
`create/environmentPlugins` style feels natural to me. Also now that name is `environmentPlugins`, it's more explicit and I like it better.

Reply:
Btw, I cannot make my comment resolved on Vite repo. Anyone who can, please feel free to resolve it anytime. Thanks!